File: management-frontend/src/utils/shipmentMonitorLines.ts
import type { Shipment, ShipmentItem, WholesaleOrderItem } from '../types';
import { productDisplayName } from './productDisplay';
import { effectiveShipmentItemQty } from './wholesaleOrderAssignment';
import { orderItemExpectedBoxes, shipmentExpectedBoxes } from './shipmentExpectedBoxes';

export type MonitorLine = {
  key: string;
  name: string;
  qty: number;
  boxes: number | null;
  imageUrl?: string;
};

function lineFromShipmentItem(
  si: ShipmentItem,
  lang: string,
  itemFallback: (id: number) => string,
): MonitorLine {
  const product = si.wholesale_order_item?.product;
  const name = product
    ? productDisplayName(product, lang)
    : itemFallback(si.wholesale_order_item_id);
  const boxes = shipmentExpectedBoxes(si);
  return {
    key: `si-${si.id}`,
    name,
    qty: effectiveShipmentItemQty(si),
    boxes: boxes > 0 ? boxes : null,
    imageUrl: product?.image_url,
  };
}

function lineFromOrderItem(oi: WholesaleOrderItem, lang: string): MonitorLine {
  const product = oi.product;
  const name = product ? productDisplayName(product, lang) : `Item #${oi.id}`;
  const boxes = orderItemExpectedBoxes(oi);
  return {
    key: `oi-${oi.id}`,
    name,
    qty: oi.quantity,
    boxes: boxes > 0 ? boxes : null,
    imageUrl: product?.image_url,
  };
}

/** Lines to show on packing monitor cards; falls back to order items when shipment lines are missing. */
export function monitorLinesForShipment(
  shipment: Shipment,
  lang: string,
  itemFallback: (id: number) => string,
): MonitorLine[] {
  const items = shipment.items ?? [];
  if (items.length > 0) {
    return items.map((si) => lineFromShipmentItem(si, lang, itemFallback));
  }

  const orderItems = shipment.wholesale_order?.items ?? [];
  if (orderItems.length === 0) return [];

  const storeId = shipment.store_id;
  const forStore = orderItems.filter(
    (oi) => oi.assigned_store_id == null || oi.assigned_store_id === storeId,
  );
  return (forStore.length > 0 ? forStore : orderItems).map((oi) => lineFromOrderItem(oi, lang));
}

File: frontend/lib/utils/shipment_monitor_lines.dart
import 'product_display.dart';
import 'shipment_expected_boxes.dart';
import 'wholesale_order_assignment.dart';

class MonitorLine {
  MonitorLine({
    required this.key,
    required this.name,
    required this.qty,
    required this.boxes,
    this.imageUrl,
  });

  final String key;
  final String name;
  final double qty;
  final double? boxes;
  final String? imageUrl;
}

MonitorLine _lineFromShipmentItem(
  Map<String, dynamic> si,
  String lang,
  String Function(int id) itemFallback,
) {
  final woItem = si['wholesale_order_item'] as Map<String, dynamic>?;
  final product = woItem?['product'] as Map<String, dynamic>?;
  final itemId = (si['wholesale_order_item_id'] as num?)?.toInt() ?? 0;
  final name = product != null ? productDisplayName(product, lang) : itemFallback(itemId);
  final boxes = shipmentExpectedBoxes(si);
  return MonitorLine(
    key: 'si-${si['id']}',
    name: name,
    qty: effectiveShipmentItemQty(si),
    boxes: boxes > 0 ? boxes : null,
    imageUrl: product?['image_url']?.toString(),
  );
}

MonitorLine _lineFromOrderItem(Map<String, dynamic> oi, String lang) {
  final product = oi['product'] as Map<String, dynamic>?;
  final id = (oi['id'] as num?)?.toInt() ?? 0;
  final name = product != null ? productDisplayName(product, lang) : 'Item #$id';
  final boxes = orderItemExpectedBoxes(oi);
  return MonitorLine(
    key: 'oi-$id',
    name: name,
    qty: (oi['quantity'] as num?)?.toDouble() ?? 0,
    boxes: boxes > 0 ? boxes : null,
    imageUrl: product?['image_url']?.toString(),
  );
}

List<MonitorLine> monitorLinesForShipment(
  Map<String, dynamic> shipment,
  String lang,
  String Function(int id) itemFallback,
) {
  final items = shipment['items'] as List<dynamic>? ?? [];
  if (items.isNotEmpty) {
    return items
        .whereType<Map<String, dynamic>>()
        .map((si) => _lineFromShipmentItem(si, lang, itemFallback))
        .toList();
  }

  final order = shipment['wholesale_order'] as Map<String, dynamic>?;
  final orderItems = order?['items'] as List<dynamic>? ?? [];
  if (orderItems.isEmpty) return [];

  final storeId = (shipment['store_id'] as num?)?.toInt();
  final forStore = orderItems.whereType<Map<String, dynamic>>().where((oi) {
    final assignedStore = (oi['assigned_store_id'] as num?)?.toInt();
    return assignedStore == null || assignedStore == storeId;
  }).toList();
  final source = forStore.isNotEmpty
      ? forStore
      : orderItems.whereType<Map<String, dynamic>>().toList();
  return source.map((oi) => _lineFromOrderItem(oi, lang)).toList();
}

String formatQtyLabel(double qty) {
  return qty == qty.roundToDouble() ? '${qty.round()}' : qty.toStringAsFixed(2);
}
